seq/src/main.rs

```diff
@@ -0,0 +1,201 @@
+use std::process;
+
+mod cli;
+
+fn main() {
+    let matches = cli::create_app().get_matches();
+
+    if let Some(values) = matches.values_of("FIRST INCREMENT LAST") {
+        let args: Vec<&str> = values.collect();
+        if args.len() > 3 {
+            eprintln!("seq: extra operand '{}'\n Try 'seq --help' for more information.", args[3]);
+            process::exit(1);
+        }
+        let seperator = matches.value_of("SEPERATOR").map(String::from).unwrap();
+        let decimals = max_decimal_digits(&args);
+        let padding = if matches.is_present("WIDTH") { Some(max_digits(&args)) } else { None };
+        let (first, inc, last) = find_operands(&args);
+        let valid_range = (first < last && inc > 0.0) || (first > last && inc < 0.0);
+        if valid_range {
+            let seq = Seq::new(first, inc, last, decimals, seperator, padding);
+            for val in seq.into_iter() {
+                print!("{}", val);
+            }
+            println!("");
+        }
+    } else {
+        eprintln!("seq: missing operand\n Try 'seq --help' for more information.");
+        process::exit(1);
+    }
+}
+
+fn find_operands(args: &[&str]) -> (f64, f64, f64) {
+    match args.len() {
+        1 => (1.0, 1.0, args[0].parse::<f64>().unwrap()),
+        2 => (args[0].parse().unwrap(), 1.0, args[1].parse().unwrap()),
+        _ => (args[0].parse().unwrap(), args[1].parse().unwrap(), args[2].parse().unwrap()),
+    }
+}
+
+struct Seq {
+    first:     f64,
+    inc:       f64,
+    last:      f64,
+    decimals:  usize,
+    seperator: String,
+    padding:   Option<usize>,
+}
+
+impl Seq {
+    fn new(
+        first: f64, inc: f64, last: f64, decimals: usize, seperator: String, padding: Option<usize>,
+    ) -> Seq {
+        Seq { first, inc, last, decimals, seperator, padding }
+    }
+
+    fn is_complete(&self, value: f64) -> bool {
+        self.inc > 0.0 && value > self.last || self.inc < 0.0 && value < self.last
+    }
+}
+
+fn max_decimal_digits(args: &[&str]) -> usize {
+    args.iter().map(|v| v.len() - v.find(".").map(|pos| pos + 1).unwrap_or(v.len())).max().unwrap()
+}
+
+fn max_digits(args: &[&str]) -> usize {
+    args.iter().map(|v| v.find(".").unwrap_or(v.len())).max().unwrap()
+}
```

Comment:
Are this unwrap never fail? If they never fail, can you add comments for why it don't fail? If they can  fail, it should pass the information forward or be handled on the site

Reply:
That's right, it should never fail. I add a comment to make it explicit.